s3/src/test/scala/uk/gov/nationalarchives/DAS3ClientTest.scala

```diff
@@ -223,6 +215,60 @@ class DAS3ClientTest extends AnyFlatSpec with MockitoSugar {
     ex.getMessage should equal("Error calling delete objects")
   }
 
+  "listKeysWithCommonPrefixes" should "return a publisher with the expected type" in {
+    // You can't mock/spy on the publisher properly due to the commonPrefix method being final,
```

Comment:
I think I got confused when you were talking about final methods before. This isn't really testing anything that the compiler won't already tell you. You can use the StepVerifier in the same way as the download tests to check for this:

```scala
  "listKeysWithCommonPrefixes" should "return a publisher with the expected common prefixes" in {
   val asyncClientMock = mock[S3AsyncClient]

    val listObjectsV2Request = ListObjectsV2Request.builder
      .bucket("bucket")
      .delimiter("/")
      .prefix("prefix")
      .build

    val response = ListObjectsV2Response.builder.commonPrefixes(CommonPrefix.builder.prefix("commonPrefix").build).build()
    val listObjectsV2Publisher = new ListObjectsV2Publisher(asyncClientMock, listObjectsV2Request)

    when(asyncClientMock.listObjectsV2(any[ListObjectsV2Request])).thenReturn(CompletableFuture.completedFuture(response))
    when(asyncClientMock.listObjectsV2Paginator(any[ListObjectsV2Request])).thenReturn(listObjectsV2Publisher)

    val client = DAS3Client[IO](asyncClientMock)
    val publisher: SdkPublisher[String] = client.listKeysWithCommonPrefixes("bucket", "keyPrefix/").unsafeRunSync()
    StepVerifier
      .create(publisher)
      .expectNext("commonPrefix")
      .verifyComplete()
  }

```

Reply:
Yes, I tried this but I get a `NullPointerException`

```
You have a NullPointerException here:
-> at software.amazon.awssdk.core.internal.pagination.async.ItemsSubscription.fetchNextPage(ItemsSubscription.java:82)
because this method call was *not* stubbed correctly:
```

Reply:
Cool. I think you can get rid of the comment at the top of the test now as well.